File: curriculum/11_load_balancer/node-impl/src/load-balancer.ts
import crypto from 'node:crypto';
import http, { IncomingMessage, Server, ServerResponse } from 'node:http';
import { URL } from 'node:url';
import httpProxy from 'http-proxy';

export type RoutingAlgorithm = 'round_robin' | 'least_connections';
export type HealthState = 'healthy' | 'unhealthy' | 'unknown';
export type CircuitState = 'closed' | 'open' | 'half_open';

export interface BackendConfig {
  id: string;
  url: string;
  weight?: number;
  healthPath?: string;
  maxConnections?: number;
}

export interface LoadBalancerConfig {
  listenAddress?: string;
  routingAlgorithm: RoutingAlgorithm;
  healthCheckIntervalMs: number;
  healthCheckTimeoutMs: number;
  healthyThreshold: number;
  unhealthyThreshold: number;
  failureThreshold: number;
  openDurationMs: number;
  backends: BackendConfig[];
}

export interface BackendSnapshot {
  id: string;
  url: string;
  weight: number;
  health: HealthState;
  activeConnections: number;
  totalRequests: number;
  failedRequests: number;
  circuitState: CircuitState;
}

interface BackendRuntime extends BackendSnapshot {
  healthPath: string;
  consecutiveSuccesses: number;
  consecutiveFailures: number;
  openedAt: number;
  halfOpenProbeInFlight: boolean;
}

export class LoadBalancer {
  private readonly backends = new Map<string, BackendRuntime>();
  private readonly proxy = httpProxy.createProxyServer({ changeOrigin: true, xfwd: false, proxyTimeout: 5000 });
  private roundRobinCursor = 0;
  private timer: NodeJS.Timeout | null = null;
  private server: Server | null = null;
  private readonly startedAt = Date.now();
  private readonly metrics = { requestsTotal: 0, requestsInFlight: 0, responsesByStatusClass: new Map<string, number>(), backendRequests: new Map<string, number>() };

  constructor(private readonly config: LoadBalancerConfig, private readonly logger: Pick<Console, 'log' | 'error'> = console) {
    if (config.backends.length === 0) throw new Error('at least one backend is required');
    if (config.healthCheckIntervalMs < 10) throw new Error('healthCheckIntervalMs must be >= 10');
    for (const backend of config.backends) this.addBackend(backend);
    this.proxy.on('proxyReq', (proxyReq, req) => {
      const requestId = req.headers['x-request-id']?.toString() ?? crypto.randomUUID();
      proxyReq.setHeader('x-request-id', requestId);
      proxyReq.setHeader('x-forwarded-host', req.headers.host ?? '');
      proxyReq.setHeader('x-forwarded-proto', 'http');
      proxyReq.setHeader('x-forwarded-for', appendForwardedFor(req));
    });
  }

  addBackend(config: BackendConfig): void {
    if (!config.id.trim()) throw new Error('backend id is required');
    if (this.backends.has(config.id)) throw new Error(`duplicate backend id ${config.id}`);
    const parsed = new URL(config.url);
    if (parsed.protocol !== 'http:' && parsed.protocol !== 'https:') throw new Error('backend url must be http(s)');
    const weight = config.weight ?? 1;
    if (!Number.isInteger(weight) || weight < 1) throw new Error('backend weight must be >= 1');
    this.backends.set(config.id, { id: config.id, url: parsed.toString().replace(/\/$/, ''), weight, healthPath: config.healthPath ?? '/health', health: 'unknown', activeConnections: 0, totalRequests: 0, failedRequests: 0, circuitState: 'closed', consecutiveSuccesses: 0, consecutiveFailures: 0, openedAt: 0, halfOpenProbeInFlight: false });
  }

  removeBackend(id: string): boolean { return this.backends.delete(id); }
  snapshots(): BackendSnapshot[] { return [...this.backends.values()].map(({ id, url, weight, health, activeConnections, totalRequests, failedRequests, circuitState }) => ({ id, url, weight, health, activeConnections, totalRequests, failedRequests, circuitState })); }
  metricsSnapshot() { return { requestsTotal: this.metrics.requestsTotal, requestsInFlight: this.metrics.requestsInFlight, responsesByStatusClass: Object.fromEntries(this.metrics.responsesByStatusClass), backendRequests: Object.fromEntries(this.metrics.backendRequests), routingAlgorithm: this.config.routingAlgorithm }; }

  selectBackend(): BackendSnapshot | null {
    const eligible = this.weightedEligible();
    if (eligible.length === 0) return null;
    if (this.config.routingAlgorithm === 'least_connections') {
      return eligible.reduce((best, candidate) => candidate.activeConnections < best.activeConnections || (candidate.activeConnections === best.activeConnections && candidate.id < best.id) ? candidate : best);
    }
    const selected = eligible[this.roundRobinCursor % eligible.length];
    this.roundRobinCursor += 1;
    return selected;
  }

  async checkBackend(id: string): Promise<void> {
    const backend = this.backends.get(id);
    if (!backend) return;
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), this.config.healthCheckTimeoutMs);
    const started = Date.now();
    try {
      const response = await fetch(`${backend.url}${backend.healthPath}`, { signal: controller.signal });
      if (response.status >= 200 && response.status < 300) this.markHealthSuccess(backend, Date.now() - started);
      else this.markHealthFailure(backend, `status_${response.status}`);
    } catch (error) {
      this.markHealthFailure(backend, error instanceof Error ? error.message : 'health_check_failed');
    } finally {
      clearTimeout(timeout);
    }
  }

  startHealthChecks(): void {
    if (this.timer) return;
    this.timer = setInterval(() => { void Promise.all([...this.backends.keys()].map((id) => this.checkBackend(id))); }, this.config.healthCheckIntervalMs);
  }

  stopHealthChecks(): void { if (this.timer) clearInterval(this.timer); this.timer = null; }

  markBackendHealthy(id: string): void { const b = this.backends.get(id); if (b) b.health = 'healthy'; }
  recordFailure(id: string): void {
    const backend = this.backends.get(id);
    if (!backend) return;
    backend.failedRequests += 1;
    backend.consecutiveFailures += 1;
    if (backend.consecutiveFailures >= this.config.failureThreshold) {
      backend.circuitState = 'open';
      backend.openedAt = Date.now();
      backend.halfOpenProbeInFlight = false;
      this.logger.error(JSON.stringify({ event: 'circuit_opened', backendId: id }));
    }
  }

  async handler(req: IncomingMessage, res: ServerResponse): Promise<void> {
    if (req.url?.startsWith('/__lb/')) { this.admin(req, res); return; }
    const selected = this.selectBackend();
    if (!selected) { writeJson(res, 503, { error: { code: 'no_eligible_backend' } }); return; }
    const backend = this.backends.get(selected.id);
    if (!backend) { writeJson(res, 503, { error: { code: 'no_eligible_backend' } }); return; }
    backend.activeConnections += 1;
    backend.totalRequests += 1;
    this.metrics.requestsTotal += 1;
    this.metrics.requestsInFlight += 1;
    this.metrics.backendRequests.set(backend.id, (this.metrics.backendRequests.get(backend.id) ?? 0) + 1);
    this.logger.log(JSON.stringify({ event: 'proxy_request', backendId: backend.id, method: req.method, url: req.url }));
    await new Promise<void>((resolve) => {
      let settled = false;
      const finish = (failed: boolean, statusCode?: number): void => {
        if (settled) return;
        settled = true;
        backend.activeConnections -= 1;
        this.metrics.requestsInFlight -= 1;
        if (statusCode) this.countStatus(statusCode);
        if (failed) this.recordFailure(backend.id); else this.recordSuccess(backend.id);
        resolve();
      };
      res.on('finish', () => finish(res.statusCode >= 500, res.statusCode));
      this.proxy.web(req, res, { target: backend.url }, (error) => {
        if (!res.headersSent) writeJson(res, error.message.includes('timeout') ? 504 : 502, { error: { code: 'bad_gateway', message: error.message } });
        finish(true, res.statusCode);
      });
    });
  }

  listen(port: number, host = '127.0.0.1'): Server {
    this.startHealthChecks();
    this.server = http.createServer((req, res) => { void this.handler(req, res); });
    this.server.listen(port, host);
    return this.server;
  }

  async shutdown(timeoutMs = 5000): Promise<void> {
    this.stopHealthChecks();
    const server = this.server;
    if (!server) return;
    await Promise.race([new Promise<void>((resolve, reject) => server.close((err) => err ? reject(err) : resolve())), new Promise<void>((resolve) => setTimeout(resolve, timeoutMs))]);
    this.proxy.close();
  }

  private weightedEligible(): BackendRuntime[] {
    const now = Date.now();
    const out: BackendRuntime[] = [];
    for (const backend of this.backends.values()) {
      if (backend.circuitState === 'open' && now - backend.openedAt >= this.config.openDurationMs) backend.circuitState = 'half_open';
      const allowedCircuit = backend.circuitState === 'closed' || (backend.circuitState === 'half_open' && !backend.halfOpenProbeInFlight);
      if (backend.health !== 'unhealthy' && allowedCircuit) for (let i = 0; i < backend.weight; i += 1) out.push(backend);
    }
    return out;
  }
  private markHealthSuccess(backend: BackendRuntime, latencyMs: number): void {
    backend.consecutiveSuccesses += 1; backend.consecutiveFailures = 0;
    if (backend.consecutiveSuccesses >= this.config.healthyThreshold) backend.health = 'healthy';
    if (backend.circuitState === 'half_open') backend.circuitState = 'closed';
    this.logger.log(JSON.stringify({ event: 'health_success', backendId: backend.id, latencyMs }));
  }
  private markHealthFailure(backend: BackendRuntime, reason: string): void {
    backend.consecutiveFailures += 1; backend.consecutiveSuccesses = 0;
    if (backend.consecutiveFailures >= this.config.unhealthyThreshold) backend.health = 'unhealthy';
    this.recordFailure(backend.id);
    this.logger.error(JSON.stringify({ event: 'health_failure', backendId: backend.id, reason }));
  }
  private recordSuccess(id: string): void {
    const backend = this.backends.get(id);
    if (!backend) return;
    backend.consecutiveFailures = 0;
    if (backend.circuitState === 'half_open') backend.circuitState = 'closed';
  }
  private countStatus(status: number): void { const key = `${Math.floor(status / 100)}xx`; this.metrics.responsesByStatusClass.set(key, (this.metrics.responsesByStatusClass.get(key) ?? 0) + 1); }
  private admin(req: IncomingMessage, res: ServerResponse): void {
    if (req.url === '/__lb/health') { const snapshots = this.snapshots(); writeJson(res, 200, { status: 'ok', uptimeSeconds: Math.floor((Date.now() - this.startedAt) / 1000), backendSummary: { healthy: snapshots.filter((b) => b.health === 'healthy').length, unhealthy: snapshots.filter((b) => b.health === 'unhealthy').length, openCircuits: snapshots.filter((b) => b.circuitState === 'open').length } }); return; }
    if (req.url === '/__lb/backends') { writeJson(res, 200, { items: this.snapshots() }); return; }
    if (req.url === '/__lb/metrics') { writeJson(res, 200, this.metricsSnapshot()); return; }
    writeJson(res, 404, { error: { code: 'not_found' } });
  }
}

function appendForwardedFor(req: IncomingMessage): string {
  const prior = req.headers['x-forwarded-for']?.toString();
  const remote = req.socket.remoteAddress ?? '';
  return prior ? `${prior}, ${remote}` : remote;
}

function writeJson(res: ServerResponse, status: number, body: unknown): void {
  res.statusCode = status;
  res.setHeader('content-type', 'application/json');
  res.end(JSON.stringify(body));
}

export const defaultConfig = (backends: BackendConfig[]): LoadBalancerConfig => ({ routingAlgorithm: 'round_robin', healthCheckIntervalMs: 1000, healthCheckTimeoutMs: 500, healthyThreshold: 1, unhealthyThreshold: 1, failureThreshold: 2, openDurationMs: 1000, backends });

File: curriculum/11_load_balancer/go-impl/loadbalancer.go
package loadbalancer

import (
	"context"
	"encoding/json"
	"errors"
	"log/slog"
	"net"
	"net/http"
	"net/http/httputil"
	"net/url"
	"strings"
	"sync"
	"time"
)

type RoutingAlgorithm string
type HealthState string
type CircuitState string

const (
	RoundRobin       RoutingAlgorithm = "round_robin"
	LeastConnections RoutingAlgorithm = "least_connections"
	Healthy          HealthState      = "healthy"
	Unhealthy        HealthState      = "unhealthy"
	Unknown          HealthState      = "unknown"
	Closed           CircuitState     = "closed"
	Open             CircuitState     = "open"
	HalfOpen         CircuitState     = "half_open"
)

type BackendConfig struct {
	ID         string
	URL        string
	Weight     int
	HealthPath string
}

type Config struct {
	RoutingAlgorithm    RoutingAlgorithm
	HealthCheckInterval time.Duration
	HealthCheckTimeout  time.Duration
	HealthyThreshold    int
	UnhealthyThreshold  int
	FailureThreshold    int
	OpenDuration        time.Duration
	Backends            []BackendConfig
}

type BackendSnapshot struct {
	ID                string       `json:"id"`
	URL               string       `json:"url"`
	Weight            int          `json:"weight"`
	Health            HealthState  `json:"health"`
	CircuitState      CircuitState `json:"circuitState"`
	ActiveConnections int          `json:"activeConnections"`
	TotalRequests     int64        `json:"totalRequests"`
	FailedRequests    int64        `json:"failedRequests"`
}

type backendRuntime struct {
	BackendSnapshot
	parsed               *url.URL
	healthPath           string
	consecutiveSuccesses int
	consecutiveFailures  int
	openedAt             time.Time
	halfOpenProbe        bool
}

type Metrics struct {
	RequestsTotal          int64            `json:"requestsTotal"`
	RequestsInFlight       int              `json:"requestsInFlight"`
	ResponsesByStatusClass map[string]int64 `json:"responsesByStatusClass"`
	BackendRequests        map[string]int64 `json:"backendRequests"`
	RoutingAlgorithm       RoutingAlgorithm `json:"routingAlgorithm"`
}

type LoadBalancer struct {
	mu        sync.RWMutex
	backends  map[string]*backendRuntime
	order     []string
	cursor    int
	config    Config
	client    *http.Client
	logger    *slog.Logger
	metrics   Metrics
	startedAt time.Time
	stop      chan struct{}
	once      sync.Once
}

func DefaultConfig(backends []BackendConfig) Config {
	return Config{RoutingAlgorithm: RoundRobin, HealthCheckInterval: time.Second, HealthCheckTimeout: 500 * time.Millisecond, HealthyThreshold: 1, UnhealthyThreshold: 1, FailureThreshold: 2, OpenDuration: time.Second, Backends: backends}
}

func New(cfg Config) (*LoadBalancer, error) {
	if len(cfg.Backends) == 0 {
		return nil, errors.New("at least one backend is required")
	}
	if cfg.RoutingAlgorithm == "" {
		cfg.RoutingAlgorithm = RoundRobin
	}
	if cfg.HealthCheckInterval <= 0 {
		cfg.HealthCheckInterval = time.Second
	}
	if cfg.HealthCheckTimeout <= 0 {
		cfg.HealthCheckTimeout = 500 * time.Millisecond
	}
	if cfg.HealthyThreshold <= 0 {
		cfg.HealthyThreshold = 1
	}
	if cfg.UnhealthyThreshold <= 0 {
		cfg.UnhealthyThreshold = 1
	}
	if cfg.FailureThreshold <= 0 {
		cfg.FailureThreshold = 2
	}
	if cfg.OpenDuration <= 0 {
		cfg.OpenDuration = time.Second
	}
	lb := &LoadBalancer{backends: map[string]*backendRuntime{}, config: cfg, client: &http.Client{Timeout: cfg.HealthCheckTimeout}, logger: slog.Default(), startedAt: time.Now(), stop: make(chan struct{}), metrics: Metrics{ResponsesByStatusClass: map[string]int64{}, BackendRequests: map[string]int64{}, RoutingAlgorithm: cfg.RoutingAlgorithm}}
	for _, backend := range cfg.Backends {
		if err := lb.AddBackend(backend); err != nil {
			return nil, err
		}
	}
	return lb, nil
}

func (lb *LoadBalancer) AddBackend(cfg BackendConfig) error {
	if strings.TrimSpace(cfg.ID) == "" {
		return errors.New("backend id is required")
	}
	if cfg.Weight == 0 {
		cfg.Weight = 1
	}
	if cfg.Weight < 0 {
		return errors.New("backend weight must be >= 1")
	}
	parsed, err := url.Parse(cfg.URL)
	if err != nil || (parsed.Scheme != "http" && parsed.Scheme != "https") {
		return errors.New("backend url must be http(s)")
	}
	if cfg.HealthPath == "" {
		cfg.HealthPath = "/health"
	}
	lb.mu.Lock()
	defer lb.mu.Unlock()
	if _, exists := lb.backends[cfg.ID]; exists {
		return errors.New("duplicate backend id")
	}
	lb.backends[cfg.ID] = &backendRuntime{BackendSnapshot: BackendSnapshot{ID: cfg.ID, URL: strings.TrimRight(parsed.String(), "/"), Weight: cfg.Weight, Health: Unknown, CircuitState: Closed}, parsed: parsed, healthPath: cfg.HealthPath}
	lb.order = append(lb.order, cfg.ID)
	return nil
}

func (lb *LoadBalancer) RemoveBackend(id string) bool {
	lb.mu.Lock()
	defer lb.mu.Unlock()
	if _, ok := lb.backends[id]; !ok {
		return false
	}
	delete(lb.backends, id)
	for i, existing := range lb.order {
		if existing == id {
			lb.order = append(lb.order[:i], lb.order[i+1:]...)
			break
		}
	}
	return true
}

func (lb *LoadBalancer) Snapshots() []BackendSnapshot {
	lb.mu.RLock()
	defer lb.mu.RUnlock()
	out := make([]BackendSnapshot, 0, len(lb.backends))
	for _, id := range lb.order {
		if b := lb.backends[id]; b != nil {
			out = append(out, b.BackendSnapshot)
		}
	}
	return out
}

func (lb *LoadBalancer) Metrics() Metrics {
	lb.mu.RLock()
	defer lb.mu.RUnlock()
	copyMetrics := Metrics{RequestsTotal: lb.metrics.RequestsTotal, RequestsInFlight: lb.metrics.RequestsInFlight, ResponsesByStatusClass: map[string]int64{}, BackendRequests: map[string]int64{}, RoutingAlgorithm: lb.metrics.RoutingAlgorithm}
	for k, v := range lb.metrics.ResponsesByStatusClass {
		copyMetrics.ResponsesByStatusClass[k] = v
	}
	for k, v := range lb.metrics.BackendRequests {
		copyMetrics.BackendRequests[k] = v
	}
	return copyMetrics
}

func (lb *LoadBalancer) MarkHealthy(id string) {
	lb.mu.Lock()
	defer lb.mu.Unlock()
	if b := lb.backends[id]; b != nil {
		b.Health = Healthy
	}
}

func (lb *LoadBalancer) SelectBackend() (BackendSnapshot, bool) {
	lb.mu.Lock()
	defer lb.mu.Unlock()
	eligible := lb.weightedEligibleLocked(time.Now())
	if len(eligible) == 0 {
		return BackendSnapshot{}, false
	}
	if lb.config.RoutingAlgorithm == LeastConnections {
		best := eligible[0]
		for _, b := range eligible[1:] {
			if b.ActiveConnections < best.ActiveConnections || (b.ActiveConnections == best.ActiveConnections && b.ID < best.ID) {
				best = b
			}
		}
		return best.BackendSnapshot, true
	}
	selected := eligible[lb.cursor%len(eligible)]
	lb.cursor++
	return selected.BackendSnapshot, true
}

func (lb *LoadBalancer) weightedEligibleLocked(now time.Time) []*backendRuntime {
	out := []*backendRuntime{}
	for _, id := range lb.order {
		b := lb.backends[id]
		if b.CircuitState == Open && now.Sub(b.openedAt) >= lb.config.OpenDuration {
			b.CircuitState = HalfOpen
			b.halfOpenProbe = false
		}
		allowedCircuit := b.CircuitState == Closed || (b.CircuitState == HalfOpen && !b.halfOpenProbe)
		if b.Health != Unhealthy && allowedCircuit {
			for i := 0; i < b.Weight; i++ {
				out = append(out, b)
			}
		}
	}
	return out
}

func (lb *LoadBalancer) CheckBackend(ctx context.Context, id string) {
	lb.mu.RLock()
	b := lb.backends[id]
	lb.mu.RUnlock()
	if b == nil {
		return
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, b.URL+b.healthPath, nil)
	if err != nil {
		lb.markFailure(id)
		return
	}
	started := time.Now()
	resp, err := lb.client.Do(req)
	if err != nil {
		lb.markHealthFailure(id)
		return
	}
	_ = resp.Body.Close()
	if resp.StatusCode >= 200 && resp.StatusCode < 300 {
		lb.markHealthSuccess(id, time.Since(started))
	} else {
		lb.markHealthFailure(id)
	}
}

func (lb *LoadBalancer) StartHealthChecks() {
	go func() {
		ticker := time.NewTicker(lb.config.HealthCheckInterval)
		defer ticker.Stop()
		for {
			select {
			case <-ticker.C:
				for _, s := range lb.Snapshots() {
					go lb.CheckBackend(context.Background(), s.ID)
				}
			case <-lb.stop:
				return
			}
		}
	}()
}
func (lb *LoadBalancer) Shutdown(context.Context) error {
	lb.once.Do(func() { close(lb.stop) })
	lb.logger.Info("load balancer shutdown")
	return nil
}

func (lb *LoadBalancer) markHealthSuccess(id string, latency time.Duration) {
	lb.mu.Lock()
	defer lb.mu.Unlock()
	if b := lb.backends[id]; b != nil {
		b.consecutiveSuccesses++
		b.consecutiveFailures = 0
		if b.consecutiveSuccesses >= lb.config.HealthyThreshold {
			b.Health = Healthy
		}
		if b.CircuitState == HalfOpen {
			b.CircuitState = Closed
		}
		lb.logger.Info("health_success", "backend", id, "latency_ms", latency.Milliseconds())
	}
}
func (lb *LoadBalancer) markHealthFailure(id string) {
	lb.mu.Lock()
	defer lb.mu.Unlock()
	if b := lb.backends[id]; b != nil {
		b.consecutiveFailures++
		b.consecutiveSuccesses = 0
		if b.consecutiveFailures >= lb.config.UnhealthyThreshold {
			b.Health = Unhealthy
		}
		lb.recordFailureLocked(b)
		lb.logger.Error("health_failure", "backend", id)
	}
}
func (lb *LoadBalancer) markFailure(id string) {
	lb.mu.Lock()
	defer lb.mu.Unlock()
	if b := lb.backends[id]; b != nil {
		lb.recordFailureLocked(b)
	}
}
func (lb *LoadBalancer) recordFailureLocked(b *backendRuntime) {
	b.FailedRequests++
	b.consecutiveFailures++
	if b.consecutiveFailures >= lb.config.FailureThreshold {
		b.CircuitState = Open
		b.openedAt = time.Now()
		b.halfOpenProbe = false
		lb.logger.Error("circuit_opened", "backend", b.ID)
	}
}

func (lb *LoadBalancer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	if strings.HasPrefix(r.URL.Path, "/__lb/") {
		lb.admin(w, r)
		return
	}
	selected, ok := lb.SelectBackend()
	if !ok {
		writeJSON(w, http.StatusServiceUnavailable, map[string]any{"error": map[string]string{"code": "no_eligible_backend"}})
		return
	}
	lb.mu.Lock()
	b := lb.backends[selected.ID]
	b.ActiveConnections++
	b.TotalRequests++
	lb.metrics.RequestsTotal++
	lb.metrics.RequestsInFlight++
	lb.metrics.BackendRequests[b.ID]++
	lb.mu.Unlock()
	lb.logger.Info("proxy_request", "backend", b.ID, "method", r.Method, "path", r.URL.RequestURI())
	proxy := httputil.NewSingleHostReverseProxy(b.parsed)
	originalDirector := proxy.Director
	proxy.Director = func(req *http.Request) {
		originalDirector(req)
		req.URL.Path = singleJoiningSlash(b.parsed.Path, r.URL.Path)
		req.URL.RawQuery = r.URL.RawQuery
		req.Host = b.parsed.Host
		addForwardingHeaders(req, r)
	}
	proxy.ErrorHandler = func(rw http.ResponseWriter, req *http.Request, err error) {
		lb.markFailure(b.ID)
		writeJSON(rw, http.StatusBadGateway, map[string]any{"error": map[string]string{"code": "bad_gateway", "message": err.Error()}})
	}
	wrapped := &trackingWriter{ResponseWriter: w, status: http.StatusOK}
	proxy.ServeHTTP(wrapped, r)
	lb.mu.Lock()
	b.ActiveConnections--
	lb.metrics.RequestsInFlight--
	lb.metrics.ResponsesByStatusClass[statusClass(wrapped.status)]++
	if wrapped.status >= 500 {
		lb.recordFailureLocked(b)
	} else {
		b.consecutiveFailures = 0
		if b.CircuitState == HalfOpen {
			b.CircuitState = Closed
		}
	}
	lb.mu.Unlock()
}

func (lb *LoadBalancer) admin(w http.ResponseWriter, r *http.Request) {
	switch r.URL.Path {
	case "/__lb/health":
		snapshots := lb.Snapshots()
		summary := map[string]int{"healthy": 0, "unhealthy": 0, "openCircuits": 0}
		for _, b := range snapshots {
			if b.Health == Healthy {
				summary["healthy"]++
			}
			if b.Health == Unhealthy {
				summary["unhealthy"]++
			}
			if b.CircuitState == Open {
				summary["openCircuits"]++
			}
		}
		writeJSON(w, http.StatusOK, map[string]any{"status": "ok", "uptimeSeconds": int(time.Since(lb.startedAt).Seconds()), "backendSummary": summary})
	case "/__lb/backends":
		writeJSON(w, http.StatusOK, map[string]any{"items": lb.Snapshots()})
	case "/__lb/metrics":
		writeJSON(w, http.StatusOK, lb.Metrics())
	default:
		writeJSON(w, http.StatusNotFound, map[string]any{"error": map[string]string{"code": "not_found"}})
	}
}

type trackingWriter struct {
	http.ResponseWriter
	status int
}

func (w *trackingWriter) WriteHeader(status int) {
	w.status = status
	w.ResponseWriter.WriteHeader(status)
}

func writeJSON(w http.ResponseWriter, status int, value any) {
	w.Header().Set("content-type", "application/json")
	w.WriteHeader(status)
	_ = json.NewEncoder(w).Encode(value)
}
func statusClass(status int) string { return string(rune('0'+status/100)) + "xx" }
func singleJoiningSlash(a, b string) string {
	aslash := strings.HasSuffix(a, "/")
	bslash := strings.HasPrefix(b, "/")
	switch {
	case aslash && bslash:
		return a + b[1:]
	case !aslash && !bslash:
		return a + "/" + b
	}
	return a + b
}
func addForwardingHeaders(req *http.Request, original *http.Request) {
	requestID := original.Header.Get("x-request-id")
	if requestID == "" {
		requestID = time.Now().UTC().Format("20060102150405.000000000")
	}
	req.Header.Set("x-request-id", requestID)
	req.Header.Set("x-forwarded-host", original.Host)
	req.Header.Set("x-forwarded-proto", "http")
	prior := original.Header.Get("x-forwarded-for")
	host, _, _ := net.SplitHostPort(original.RemoteAddr)
	if prior != "" {
		req.Header.Set("x-forwarded-for", prior+", "+host)
	} else {
		req.Header.Set("x-forwarded-for", host)
	}
}
